dpgen2/entrypoint/submit.py

```diff
@@ -732,32 +733,81 @@ def print_list_steps(
 
 
 def successful_step_keys(wf):
-    all_step_keys_ = wf.query_keys_of_steps()
-    wf_info = wf.query()
     all_step_keys = []
-    for ii in all_step_keys_:
-        if wf_info.get_step(key=ii)[0]["phase"] == "Succeeded":
-            all_step_keys.append(ii)
+    for step in wf.query_step():
+        if step.key is not None and step.phase == "Succeeded":
+            all_step_keys.append(step.key)
     return all_step_keys
 
 
+def get_superop(key):
+    if "prep-train" in key:
+        return key.replace("prep-train", "prep-run-train")
+    elif "run-train" in key:
+        return re.sub("run-train-[0-9]*", "prep-run-train", key)
+    elif "prep-lmp" in key:
+        return key.replace("prep-lmp", "prep-run-explore")
+    elif "run-lmp" in key:
+        return re.sub("run-lmp-[0-9]*", "prep-run-explore", key)
+    elif "prep-fp" in key:
+        return key.replace("prep-fp", "prep-run-fp")
+    elif "run-fp" in key:
+        return re.sub("run-fp-[0-9]*", "prep-run-fp", key)
+    elif "prep-caly-input" in key:
+        return key.replace("prep-caly-input", "prep-run-explore")
+    elif "collect-run-calypso" in key:
+        return re.sub("collect-run-calypso-[0-9]*-[0-9]*", "prep-run-explore", key)
+    elif "prep-run-dp-optim" in key:
+        return re.sub("prep-run-dp-optim-[0-9]*-[0-9]*", "prep-run-explore", key)
+    return None
+
+
+def fold_keys(all_step_keys):
```

Comment:
you may want to write a UT for this function

Reply:
Done.